File: android/src/main/kotlin/com/snowplowanalytics/snowplow_flutter_tracker/readers/configurations/GlobalContextsConfigurationReader.kt
package com.snowplowanalytics.snowplow_flutter_tracker.readers.configurations

import com.snowplowanalytics.snowplow.globalcontexts.GlobalContext
import com.snowplowanalytics.snowplow_flutter_tracker.readers.events.SelfDescribingJsonReader

class GlobalContextsConfigurationReader(val values: Map<String, Any>) {
    val tag: String by values
    val globalContexts: List<Map<String, Any>> by values

    fun toGlobalContext(): GlobalContext {
        val staticContexts = globalContexts.map { globalContext ->
            SelfDescribingJsonReader(globalContext).toSelfDescribingJson()
        }
        return GlobalContext(staticContexts)
    }

}

File: ios/Classes/readers/configurations/GlobalContextsConfigurationReader.swift
//
//  GlobalContextsConfigurationReader.swift
//  snowplow_flutter_tracker
//
//  Created by Matus Tomlein on 03/01/2022.
//

import Foundation
import SnowplowTracker

struct GlobalContextsConfigurationReader: Decodable {
    let tag: String
    let globalContexts: [SelfDescribingJsonReader]
}

extension GlobalContextsConfigurationReader {
    func toGlobalContext(arguments: [String: Any]) -> GlobalContext? {
        if let gcArguments = arguments["globalContexts"] as? [[String: Any]] {
            let staticContexts = zip(globalContexts, gcArguments).map { (reader, data) in
                reader.toSelfDescribingJson(arguments: data)
            }.compactMap { $0 }

            return GlobalContext(staticContexts: staticContexts)
        }
        return nil
    }
}
